### PR TITLE
Create OrchardCore.Profile Module for FE Users

### PR DESCRIPTION
**Pull request by themattkenney**
Create OrchardCore Profile Module for FE Users. The module will show under security called user profile. Has been tested on Agency Theme and Blog Theme. Need to add Readme still

### CODE REVIEW

**Comment by Skrypt** [comment]
Awesome I need to test this 😄

**Comment by themattkenney** [comment]
Added navigation support and fixed styling on profile page on blog module

**Comment by Skrypt** [comment]
I tried last night quick but I was looking to see a menu where there was none. Do we need to create a Menu in Orchard with a specific name for that to show?

**Comment by themattkenney** [comment]
Once you turn on module the /profile url will display  a basic page. In your own module you can extend the page I'll post an example tomorrow

**Comment by themattkenney** [comment]
Example of how we are using in our advanced forms module

picturemessage_q4bwzrsh qww_20190222_225730350

**Comment by Skrypt** [comment]
Ok I understand, I did look really quick but that's really nice. I'm myself too using shapes that comes from the admin to display content items on a dashboard. As I can see, this looks like a good way to go.

**Comment by themattkenney** [comment]
Added Readme file. We are finished with any known changes on our end.

**Comment by jptissot** [comment]
I am not sure the name Profile fits for this module. I would expect a profile to have a page where the user can add additional information about themselves. There should also be an easy way to modify what data is captured for a specific user. 

For example, we could have a content type that allows us to define the fields we want to capture for the user's profile. The profile module would then allow the user to modify and view this information.

**Comment by Skrypt** [comment]
It's a frontend dashboard

**Comment by themattkenney** [comment]
At first, we matched the basic functionality from Orchard. We would like to add the ability for FE users to edit their registration info and see all information stored on them. I have some questions on how we should go about this to make it flexible. 

Instead of allowing groups access /admin this gives OC devs the ability to create a front-end suite of tools without adding items to /admin.   We currently use Profile to show front end-users all the Advanced Forms they have submitted and all cases they belong too.

**Comment by jptissot** [comment]
Yeah I think the name dashboard fits a bit better. It would be nice to be able to manage this dashboard from the admin UI.

File: src/OrchardCore.Modules/OrchardCore.Profile/Controllers/ProfileController.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            if (profile.UserName != User.Identity.Name)
+            {
+                profile.UserName = User.Identity.Name;
+                await _profileService.UpdateProfileAsync(profile);
+            }
+
+            var viewModel = new ProfileIndexViewModel
+            {
+                GroupId = groupId,
+                Shape = await _profileDisplayManager.BuildEditorAsync(profile, this, false, groupId)
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Profile/Controllers/ProfileController.cs]: Here is one exemple of usage of the DisplayManager BuildEditorAsync on frontend.
If we have components that have controllers that are marked with [admin] in that case this kind of usage should'nt work.
@sebastienros

**Comment by jtkech** [comment]
@themattkenney because in the last meeting they were talking about your PR.

So, just for infos, there have been some changes around caching and token rules, how we `Get` an object for caching at the tenant level, and in a given scope how we `Load` an object for mutation.

See in `TemplatesManager` as an example, here some of the rules

- Use `GetSomething()` to get data to be cached at the tenant level (e.g memory cache), and mark them as read only. Use `LoadSomething()` to load isolated data from the store to be updated. There is a session helper to use this pattern. And in a method that mutates data, check if not read only.

- Get a change token before getting an object for caching, then use it for the cache entry, so that, with a distributed implementation of the signal service, a distributed signal will be able to invalidate local caches. On updating use a `_signal.DeferredSignalToken()` that will be send after committing the session so that other instances receiving the signal will load the right data from a shared source.

**Comment by themattkenney** [comment]
@jtkech Thank you for the info. We probably won't be able to fix it until mid-Dec. The code you pointed me to could help with other internal modules we have developed also. Once we get this complete we will recommit. For us, this module is a crucial component for our non-admin users.

**Comment by agriffard** [comment]
@themattkenney Can you please solve the conflicts?

**Comment by themattkenney** [comment]
Yes, sorry this fail off my radar.

**Comment by MohitNaroliya** [comment]
@agriffard - I have resolved the conflicts and committed on the dev branch.

**Comment by themattkenney** [comment]
@agriffard @MohitNaroliya was able to get this updated. Sorry for the delay we let this get off our radar.

**Comment by agriffard** [comment]
Now that the owner is a user Id, should there be any changes?

**Comment by 4m-world** [comment]
@agriffard  with the latest build the owner is back to the username

**Comment by deanmarcussen** [comment]
@4m-world Not sure how this pr relates to your comment.

The `Owner` is the `UserId`, not `Username` on the latest dev builds. If you have issues with something related to this please open a seperate issue with the details.

**Comment by themattkenney** [comment]
@agriffard sorry look like I missed a few comments. Would you like us to update our code for the userid? If so, I will get it updated.

**Comment by Piedone** [comment]
I sincerely apologize for us taking so much time here. I checked out the PR, going over old ones. This would be useful, but in the meantime, we have user profiles, managed on the dashboard:

image

These are extensible too.

So, this would need to be integrated.

Is this something you'd like to revisit or should we close?

**Comment by themattkenney** [comment]
We can close for now.

**Comment by hishamco** [comment]
@themattkenney it would be nice if we have this one in [Orchard Core Contrib (OCC)](https://github.com/orchardcoreContrib/), I remember there's an opened issue https://github.com/OrchardCoreContrib/OrchardCoreContrib/issues/16 for EF Core integration, it would be great if we have a contribution if you like

**Pull request closed without merging**